File: 1721-swapping-nodes-in-a-linked-list/1721-swapping-nodes-in-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapNodes(ListNode head, int k) {
        ArrayList<Integer> nums = new ArrayList<>();

        ListNode current = head;
        while (current != null) {
            nums.add(current.val);
            current = current.next;
        }

        int temp = nums.get(k - 1);
        nums.set(k - 1, nums.get(nums.size() - k));
        nums.set(nums.size() - k, temp);

        current = head;
        for (int i = 0; i < nums.size(); ++i) {
            current.val = nums.get(i);
            current = current.next;
        }

        return head;
    }
}

File: 1721-swapping-nodes-in-a-linked-list/1721-swapping-nodes-in-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k) {
        ListNode* fast = head;
        ListNode* slow = head;

        int top = k;
        while (k-- > 0) {
            fast = fast->next;
        }

        while (fast) {
            fast = fast->next;
            slow = slow->next;
        }
        fast = head;
        while (--top > 0) fast = fast->next;


        top = fast->val;
        fast->val = slow->val;
        slow->val = top;

        return head;
        
    }
};